### PR TITLE
Extending Muon Optimizer Support for ZeRO Stage 3

### PR DESCRIPTION
**Pull request by PKUWZP**
Authors: @pengdurice and @PKUWZP

Create a separate PR based on #7798 with the same functional diff on a clean signed-off branch to resolve DCO issues.

We aim on adding Muon Optimizer to zero stage 3 in this draft PR:

- Created a dedicated momentum buffer in zero stage 3 optimizer to save the momentum buffers specifically for Muon Optimizer.
- The optimizer states can be dispatched into 3 devices: GPU, CPU and NVME. For GPU and CPU, we just make the new buffers the same device of `self.fp32_partitioned_groups_flat`; when `device == NVME`, we make sure that the momentum buffers can be swapped in and out along with other components in the optimizer states.
- The new momentum buffers are also partitioned like `self.fp32_partitioned_groups_flat` to save memory footprint. So, before the muon update, we need to perform `all_gather` on top of each data-parallel group rank. The Muon updates of the parameters are also divided across the data-parallel ranks, and the results are all-gathered once all updates are complete. After the `all_gather`, the momentum buffers are partitioned and flattened again.

Next steps:
- Explore quantization of momentum buffers for saving memory
- Explore using highly optimized Adam / AdamW Optimizers

### CODE REVIEW

File: tests/unit/ops/muon/test_muon.py [hunk 1]
```diff
@@ -42,8 +43,16 @@ def test(self, optimizer_type, zero_stage, lr, hidden_dim, nlayer):
             },
             "zero_optimization": {
                 "stage": zero_stage,
-            }
+                "reduce_scatter": False,
+            },
+            "save_muon_momentum_buffer_in_memory": True,
```
Comment (tohtana) [inline tests/unit/ops/muon/test_muon.py]: We should test both True and False for save_muon_momentum_buffer_in_memory

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -329,12 +331,15 @@ def _enforce_optimizer_offload():
         self.all2all_process_group = all2all_process_group
 
         self.reduce_scatter = reduce_scatter
-
+        self.use_muon = isinstance(self.optimizer, MuonWithAuxAdam)
+        self.save_muon_momentum_buffer_in_memory = ds_config.get('save_muon_momentum_buffer_in_memory', False)
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py]: I think it is cleaner to have have the default setting in (one of) `config.py`.

File: deepspeed/runtime/zero/stage3.py [hunk 3]
```diff
@@ -1430,6 +1470,118 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+                    self._create_momentum_buffer(self.fp16_partitioned_groups_flat_numel[i], i,
+                                                 self.fp32_partitioned_groups_flat[i].ds_id)
+                state_buffer = self.optimizer.state[self.fp32_partitioned_groups_flat[i]]["momentum_buffer"]
+                for param, dest_offset, _ in group_items:
+                    momentum_buffer.append(state_buffer.narrow(0, dest_offset, param.partition_numel()).clone())
+            elif self.save_muon_momentum_buffer_in_memory:
+                state_buffer = self.muon_momentum_buffer_partitioned_groups_flat[i]
+                for param, dest_offset, _ in group_items:
+                    momentum_buffer.append(state_buffer.narrow(0, dest_offset, param.partition_numel()).clone())
+            else:
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py:1525]: No valid path for `self._swappable_optimizer_subgroup(i) == False` and `self.save_muon_momentum_buffer_in_memory == False`.
If we disable both Muon and NVMe offloading, it will fail.

File: deepspeed/runtime/zero/stage3.py [hunk 4]
```diff
@@ -907,6 +922,20 @@ def _get_sub_group_partitions(self, sub_group_id):
...
+                                                        device=self.device,
+                                                        dtype=self.communication_data_type)
+            unpinned_fp32_buffer_momentum.requires_grad = False
+            if self.fp32_partitioned_groups_flat[i] not in self.optimizer.state:
+                self.optimizer.state[self.fp32_partitioned_groups_flat[i]] = {}
+            self.optimizer.state[
+                self.fp32_partitioned_groups_flat[i]]["momentum_buffer"] = unpinned_fp32_buffer_momentum
+            if self.save_muon_momentum_buffer_in_memory:
+                self.muon_momentum_buffer_partitioned_groups_flat.append(unpinned_fp32_buffer_momentum)
+                self.muon_momentum_buffer_partitioned_groups_flat[i].ds_id = ds_id
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py:943]: We append an element to `muon_momentum_buffer_partitioned_groups_flat` only for `muon param group (self.sub_groups_using_muon[i] is True)`, but we access `self.muon_momentum_buffer_partitioned_groups_flat[i]` where `i` is the global index.
If we have non-muon param group earlier, `i` will be beyond the boundary.

File: deepspeed/runtime/zero/stage3.py [hunk 5]
```diff
@@ -780,6 +787,14 @@ def _create_fp16_partitions_with_defragmentation(self, fp16_param_groups):
         param_groups: List[List[Parameter]] = tuple(
             self._create_fp16_sub_groups(param_group["params"]) for param_group in fp16_param_groups)
 
+        if self.use_muon:
+            self.sub_groups_using_muon = []
+            for idx, param_group in enumerate(fp16_param_groups):
+                if getattr(param_group['params'][0], 'use_muon', False):
+                    self.sub_groups_using_muon.extend([True] * len(param_groups[idx]))
+                    self.muon_beta = param_group['momentum']
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py]: If you ave multiple muon param groups. `self.muon_beta` will be overwritten.

Comment (PKUWZP) [reply deepspeed/runtime/zero/stage3.py]: @tohtana Good call out, I have fixed the issue.

File: deepspeed/runtime/zero/stage3.py [hunk 6]
```diff
@@ -1430,6 +1470,118 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+            params_size_offset += param.grad.numel()
+
+        # process muon updates per subgroup to avoid holding all parameters and states at once
+        for i, group_items in params_by_group.items():
+            params = [param for param, _, _ in group_items]
+            if not params:
+                continue
+
+            momentum_buffer = []
+            if self._swappable_optimizer_subgroup(i) and not self.save_muon_momentum_buffer_in_memory:
```
Comment (tohtana) [inline deepspeed/runtime/zero/stage3.py:1512]: The condition looks correct, but consider simplifying like the following (should be equivalent)

```
  if self.save_muon_momentum_buffer_in_memory:
      ...
  elif self._swappable_optimizer_subgroup(i):
      ...
  else:
      ...
```

**Comment by delock** [comment]
Suggest to turn on stage 3 for `test_muon_partial_training.py` as well, this would check the case when almost all parameters are freezed so all trainable parameters use muon optimizer.

**Comment by delock** [comment]
I remember deepspeed allow seperate learning rate for muon and adam (muon_lr and adam_lr), can we have a config in UT to cover this usage?

**Review by delock — APPROVED** [review]
LGTM, some test case coverage suggestion are added to comments.

File: docs/_pages/config-json.md [hunk 7]
```diff
@@ -62,6 +64,24 @@ The Adam optimizer also supports the following two params keys/values in additio
...
+      "lr": 0.001,
+      "momentum": 0.9,
+      "weight_decay": 0.0,
+      "muon_lr": 0.001
+    }
+  },
+  "zero_optimization": {
+    "stage": 3,
+    "save_muon_momentum_buffer_in_memory": true
+  }
```
Comment (delock) [inline docs/_pages/config-json.md:82]: Did Muon optimizer for stage 3 mandates `reduce_scatter = false`?   Does `"reduce_scatter": false` need to be added to the example?
